sqlglot/jsonpath.py

```diff
@@ -142,111 +143,55 @@ def _parse_bracket() -> exp.JSONPathPart:
 
         return node
 
-    nodes: t.List[exp.JSONPathPart] = []
+    # We canonicalize the JSON path AST so that it always starts with a
+    # "root" element, so paths like "field" will be generated as "$.field"
+    _match(TokenType.DOLLAR)
+    expressions: t.List[exp.JSONPathPart] = [exp.JSONPathRoot()]
 
     while _curr():
-        if _match(TokenType.DOLLAR):
-            nodes.append(exp.JSONPathRoot())
-        elif _match(TokenType.DOT) or _match(TokenType.COLON):
+        if _match(TokenType.DOT) or _match(TokenType.COLON):
             recursive = _prev().text == ".."
             value = _match(TokenType.VAR) or _match(TokenType.IDENTIFIER) or _match(TokenType.STAR)
 
             if recursive:
-                nodes.append(exp.JSONPathRecursive(this=value.text if value else None))
+                expressions.append(exp.JSONPathRecursive(this=value.text if value else None))
             elif value:
-                nodes.append(exp.JSONPathKey(this=value.text))
+                expressions.append(exp.JSONPathKey(this=value.text))
             else:
                 raise ParseError(_error("Expected key name after DOT"))
         elif _match(TokenType.L_BRACKET):
-            nodes.append(_parse_bracket())
+            expressions.append(_parse_bracket())
         elif _match(TokenType.VAR) or _match(TokenType.IDENTIFIER):
-            nodes.append(exp.JSONPathKey(this=_prev().text))
+            expressions.append(exp.JSONPathKey(this=_prev().text))
         elif _match(TokenType.STAR):
-            nodes.append(exp.JSONPathWildcard())
+            expressions.append(exp.JSONPathWildcard())
         else:
             raise ParseError(_error(f"Unexpected {tokens[i].token_type}"))
 
-    # We canonicalize the JSON path AST so that it always starts with a
-    # "root" element, so paths like "field" will be generated as "$.field"
-    if nodes and not isinstance(nodes[0], exp.JSONPathRoot):
-        nodes.insert(0, exp.JSONPathRoot())
-
-    return nodes
+    return exp.JSONPath(expressions=expressions)
 
 
-def _generate_json_path_key(node: exp.JSONPathKey, **kwargs) -> str:
-    this = node.this
-    if this == "*" or exp.SAFE_IDENTIFIER_RE.match(this):
-        return f".{this}"
+def generate(path: exp.JSONPath, dialect: DialectType = None) -> str:
+    """Generates a string from a JSONPath expression, optionally targetting a given SQL dialect."""
+    return path.sql(dialect=dialect)[1:-1]  # The slice is used to remove the string delimiters
```

Comment:
this isn't used anywhere right? this makes a copy of the expression

Reply:
May need to refactor this one yeah, it's only used in the test suite for json paths. Previously we'd generate an unquoted path and then later stringify it in the generator, but with this approach the quotes are added first and the tests fail because they weren't expected in cts.json

Reply:
Removed in https://github.com/tobymao/sqlglot/pull/2920/commits/d29aeb70b2497a9562e36492cc7a29f049214cb7

Reply:
Removed in https://github.com/tobymao/sqlglot/pull/2920/commits/d29aeb70b2497a9562e36492cc7a29f049214cb7